### ISSUE TITLE
[analyzer] add a hint for unnecessary switch case

### ISSUE THREAD

**Issue by asashour** [dart-lang/sdk#50483]
Similar to [unnecessary_null_comparison](https://dart.dev/tools/diagnostic-messages?utm_source=dartdev&utm_medium=redir&utm_id=diagcode&utm_content=unnecessary_null_comparison#unnecessary_null_comparison) and the recent `UNNECESSARY_NAN_COMPARISON`, we should have hints for the switch case, e.g.:

```dart
void f(int i) {
  switch(i) {
    case null:
      break;
  }
}
```

and
```dart
void f(double d) {
  switch (d) {
    case double.nan:
      break;
  }
}
```

**Comment by bwilkerson** [comment]
The second case might be useful, though I doubt that it occurs frequently enough in real code to make it worth checking.

But I'm confused by the first case. @eernstg 

The current version of the spec reads

> It is a compile-time error if the type of _e_ may not be assigned to the type of _e<sub>k</sub>_.

In the first example, the type of _e_ is _int_ and the type of _e<sub>k</sub>_ is _Null_, and I believe that _int_ can't be assigned to _Null_. So I think that there should already be a diagnostic here. Is the analyzer wrong, or did I confuse myself?

**Comment by eernstg** [comment]
The case about `double.nan` is mentioned in https://github.com/dart-lang/language/issues/2587. (It used to be a compile-time error to have `case double.nan:`, so the topic only comes up when we assume that Dart already includes support for patterns, and that issue is about patterns and doubles). The issue https://github.com/dart-lang/language/issues/2587 has been closed, but similar topics could appear elsewhere.

Issue https://github.com/dart-lang/language/issues/2632 contains a discussion about impossible matches (in particular, trying to match records that are statically known to be different), and `case double.nan:` is yet another impossible match. I mentioned it [there](https://github.com/dart-lang/language/issues/2632#issuecomment-1318453472).

From the language specification:
> It is a compile-time error if the type of _e_ may not be assigned to the type of _e<sub>k</sub>_.

The updates to the language specification concerning null safety have not yet been integrated into the language specification, so we need to look [here](https://github.com/dart-lang/language/blob/master/accepted/2.12/nnbd/feature-specification.md#:~:text=given%20a%20switch%20statement). But it's still a compile-time error to have a scrutinee of type `int` and a case expression of type `Null` (because we do not have `Null <: int`).

The tricky part is that we could have the function `f` in a null safe library, and it could be executed as part of a program containing legacy code as well. In this case the execution would have to use the mode 'no-sound-null-safety', and it would then be possible for `i` to have the value null. But that's a transition problem, it will go away when all code is null safe.

**Comment by bwilkerson** [comment]
So we should probably hold off on trying to catch `case double.nan` until we see what patterns will do to the spec in this regard.

But I'm not as clear about the first example. It sounds to me like there should be a compile time error for that case. (I'm assuming here that the analyzer is already correctly handling the alternate definition of assignable when one type comes from a legacy library and the other doesn't.) Is that the case?

**Comment by eernstg** [comment]
> there should be a compile time error for that case [that is, `case null:`]

Yes, I agree that there should be an error according to the rules for Dart with null safety. That is also what I get, e.g., from  DartPad ('Based on Flutter 3.3.8 Dart SDK 2.18.4'). So the analyzer and the common front end are both in good shape wrt this check.

The analyzer should not behave differently in the case where a library is intended to run with 'no-sound-null-safety' semantics. It's possible to violate soundness in that case, but the static analysis is the same (so the analyzer does not, and should not, "know" that `int i` can be null).

**Comment by asashour** [comment]
Thanks. For the first case, in DartPad `master` channel and in VS Code, there is a diagnostic.

In IntelliJ, no error. For further investigations.

**Comment by asashour** [comment]
Well, it seems there is a [diagnostic](https://dart.dev/tools/diagnostic-messages?utm_source=dartdev&utm_medium=redir&utm_id=diagcode&utm_content=case_expression_type_is_not_switch_expression_subtype#case_expression_type_is_not_switch_expression_subtype) only if the language is `2.18` (`// @dart=2.18`).

It seems to be on purpose to target only versions from that one onwards, and seems to be handled in #50502

**Comment by srawlins** [comment]
Warnings now exist for each case! 🎉
